[sw, dif_plic] Use generated register definitions

silvestrst opened the pull request:
This PR fixes #1508, and consists from two changes:
1) Generate register PLIC register definitions.
2) Change DIF PLIC to use the generated register definitions instead of the hardcoded.

The change 2) requires further tweak that I hope @eunchan can help me with. The issue is that the generated register definitions in some cases define the same definition twice. Example:
```
// Interrupt Source mode. 0: Level, 1: Edge-triggered                                                                                                                                               
#define RV_PLIC_LE0(id) (RV_PLIC##id##_BASE_ADDR + 0xc)
#define RV_PLIC_LE0_REG_OFFSET 0xc
#define RV_PLIC_LE0 0
...
// msip for Hart 0.                                                                                                                                                                                 
#define RV_PLIC_MSIP0(id) (RV_PLIC##id##_BASE_ADDR + 0x214)
#define RV_PLIC_MSIP0_REG_OFFSET 0x214
#define RV_PLIC_MSIP0 0
```
Looking at the `hw/top_earlgrey/ip/rv_plic/data/autogen/rv_plic.hjson`, if you take register IE, the `name` in `fields` has value of `E`. It seems that other registers apart from `MSIP0` and `LE` follow this convention.

It seems that when `name` in `multireg` and `name` in `fields` matches:
```
{ multireg: {
        name: "LE",
        desc: "Interrupt Source mode. 0: Level, 1: Edge-triggered",
        count: "NumSrc",
        cname: "RV_PLIC",
        swaccess: "rw",
        hwaccess: "hro",
        fields: [
          { bits: "0", name: "LE", desc: "L0E1" }
        ],
      }
    },
```
regtool produces `#define RV_PLIC_LE0` and not `#define RV_PLIC_LE0_LE0`, whether that is by design or a bug I am not sure.

imphil commented:
Are those definitions correct and usable for your purpose? I'd add a second commit to this PR where you actually use these headers instead of the handcrafted ones, but if you have local testing that shows things work, we can also merge this PR first and then you do a follow-up one with the actual rv_plic DIF changes.

silvestrst commented:
> Are those definitions correct and usable for your purpose? I'd add a second commit to this PR where you actually use these headers instead of the handcrafted ones

I will do this instead of a separate commit, I think this will also make it easier to explain the problem in the issue I am about to open.

silvestrst commented:
@eunchan could you please help me to find correct solution to fix the duplicated defines? Would it be sufficient to change `name` in `fields` to something like:
```
multireg: {
        name: "LE",
        desc: "Interrupt Source mode. 0: Level, 1: Edge-triggered",
        count: "NumSrc",
        cname: "RV_PLIC",
        swaccess: "rw",
        hwaccess: "hro",
        fields: [
          { bits: "0", name: "E", desc: "L0E1" }
        ],
      }
    },
```

Or is it a genuine bug, and the generator script needs changing?

tjaychen commented:
i took a quick peek at this (since i know @eunchan has a bunch of things to do), the issue seems to be here[https://github.com/lowRISC/opentitan/blob/master/util/reggen/gen_cheader.py#L97](url).

The problem is, when i take this out, the process still errors elsewhere, because there is an implicit assumption that when a field name is not there, it takes on the regname (so for example, `uart_rdata` field becomes `uart_rdata_rdata`).

Maybe it's worth deciding how we want this to work exactly before changing?

silvestrst commented:
@tjaychen can we not use a similar format to what `IP` and `IE` registers use?

silvestrst commented:
@tjaychen what do you think of:
[hw, plic] Change LE and MSIP register bit names

Could it go in at least for now to unblock this issue? Thanks.

silvestrst commented:
Hi @eunchan and @tjaychen , could you please have a look at this change, specifically 
4f0e8e5 .

Unfortunately this is outside the zone of my expertise, it seems fine to me, but I would like someone to double check.

eunchan commented:
Sorry for late jump-in.

The question about the bug/ or intention, i think this is intended when I looked at the code. Omitting the field name, if only one field exists, can make the code (connecting at the top) clean with a little bit less formalized. 

I think this feature creates a little bit of weirdness with multireg feature. With multireg, only the register name is visible but the register is not actual register but the field of multiple registers.

The design change `.hjson` still doesn't affect the HW structure in RTL, I think. Is this change in the `hjson` good for software? Then I am okay with this to be merged with a nit. `hw/ip/rv_plic/data/rv_plic.hjson.tpl` also should be revised as the `rv_plic.hjson` is generated file.

tjaychen commented:
@silvestrst 
sorry for not responding sooner.  I think what you're proposing makes sense to me, I just meant more on a higher level how we want this to behave in general.

I'm personally fine with always appending the field name to the reg name, even if that creates duplication and long strings.  That's just my opinion though. 

This looks good to me from my perspective.

silvestrst commented:
> Sorry for late jump-in.

> sorry for not responding sooner.

No worries guys, I understand you are busy :)

> the process still errors elsewhere, because there is an implicit assumption that when a field name is not there, it takes on the regname (so for example, uart_rdata field becomes uart_rdata_rdata).

@tjaychen I think I misunderstood what you were saying. I thought  the error was related to the register generation itself (produced header), but I guess you meant that the error occurs in consumers, like:
```
../../../sw/device/lib/dif/dif_uart.c:37:46: error: 'UART_WDATA_OFFSET' undeclared (first use in this function); did you mean 'UART_WDATA_REG_OFFSET'?
   37 |   uint32_t reg = (byte & UART_WDATA_MASK) << UART_WDATA_OFFSET;
```
As the resulting defines are now of the following form:
```
...
#define UART_WDATA_WDATA_MASK 0xff
...
```
I think your proposed solution is probably the best option. It makes the definition names a bit longer, but I guess it makes sense and consistent in terms of how the registers and fields are defined (REG_NAME + REG_FIELD + OTHER_QUALIFIERS).

I will create another pull-request, that way it will only affect software, and I think we probably can live with longer names.

Thank you @tjaychen and @eunchan

silvestrst commented:
In the latest push:
* I have rebased the change on up-to-date master
* Dropped the [hw, plic] Change LE and MSIP register bit names, as it is resolved by #1566

silvestrst commented:
I think this change is ready to go in, I have ran some tests on my consecutive IRQ test PR, didn't observe any issues to due with PLIC.
